### ISSUE TITLE
Update SpiderMonkey to Firefox v48

### ISSUE THREAD

**Issue by dumganhar** [cocos2d/cocos2d-x#16387]
Since Cocos Runtime was dead, how about updating JS VM(SpiderMonkey) to the latest version in Firefox v48.?
Currently, we're using a very very old version, firefox v32. It's a version about two years ago.

**Comment by minggo** [comment]
@pandamicro what's your opinion?

**Comment by DavidDeSimone** [comment]
My team recently investigated this issue while looking at https://github.com/cocos2d/cocos2d-x/issues/16327. We came across the following bugzilla report, https://bugzilla.mozilla.org/show_bug.cgi?id=1257185, which states that building without JIT has been broken since Spidermonkey v38. Seeing as this bug as been open since 2016-03-16 without Mozilla commenting on it, I assume it is not a priority for Mozilla. 

One of my team members attempted to get Spidermonkey building without JIT, and found it a difficult task to approach. It seems doable, however our team wasn't looking to spend the time to do it at this moment.

**Comment by minggo** [comment]
@DavidDeSimone why you need to build without JIT? As i know, JIT can be disabled in codes. Am i wrong?

**Comment by DavidDeSimone** [comment]
I would defer to @pandamicro here. When we researched disabling JIT, the only documentation and suggestions we could find on the internet (and in Mozilla's documentation) involved setting various JIT-disabling flags during the build process. These flags are currently set in the build process of the version of Spidermonkey used in cocos for iOS. 

However you mentioning disabling JIT at run time would explain the following two lines in ScriptingCore.cpp, which my team did not understand when we came across them.

```
JS::RuntimeOptionsRef(_rt).setIon(true);
JS::RuntimeOptionsRef(_rt).setBaseline(true);
```

These functions appear to enable/disable JIT (both IonMonkey and the Baseline JIT) at run time (however they are being set to true for both JIT and non-JIT platforms). I cannot find any documentation anywhere on these functions, besides references in scattered bug reports in Mozilla's bug tracker. I would appreciate any insight that pandamicro or another contributor has into this. 

My team is definitely interested in an upgraded spidermonkey, so we would appreciate any technical insight into this problem.

**Comment by pandamicro** [comment]
Disabling JIT is indeed only for iOS, because Apple has forbidden it. As we disabled it for iOS build, Runtime setting with `JS::RuntimeOptionsRef` won't have any effect on iOS, it won't trigger problem for compilation neither.

@dumganhar I totally agree with updating Spidermonkey to latest version, big benefit. But we still have a very important task to do before it:

We need a generic abstract layer of JSB API, now we are using SpiderMonkey API directly, but it's too unstable, APIs change more or less during each version update, this has affected :
1. Our own bindings code and bindings generator
2. User project which have their own bindings
3. AnySDK and SDKBOX bindings
4. Creator bindings

Currently, AnySDK has to maintain at least 6 versions of bindings code, because they need to support all versions of Cocos2d-x, this is a nightmare for their team.

For the above reason, we can not give user or our partner team more versions of bindings code, to avoid future changes in bindings API, we need to define our own script bindings API in the abstract layer.

A generic abstract layer is for making our Bindings code absolutely independent from Spidermonkey APIs. The perfect result is that it defines a generic script binding APIs, for no matter which version of Spidermonkey, no matter which script engine we are using (Spidermonkey, V8, even for Lua).

Ricardo and I have already done some work in JS new memory model research and have been merged in v3.11, but it's not complete as this wasn't the first target of our task.

So only after the abstract layer is done, we can upgrade Spidermonkey to the latest version.

@dumganhar @DavidDeSimone @minggo

**Comment by minggo** [comment]
@pandamicro sounds reasonable to me. But i think we need to a plan of it or we can not upgrade SpiderMonkey forever.

**Comment by pandamicro** [comment]
Yes, the most important thing now is to define an API set that is similar to the current one and can delegate all functionalities. 

I think we should first assign someone to design the API set, then have a discussion on it.

For reference, I know we have an user tried to do the same, they are using C++ macro to define unified types and functions for JS & Lua engine, @linshun should have some of their implementation details.

**Comment by pandamicro** [comment]
Also, we talked about [jxcore](https://github.com/jxcore/jxcore) long before, they have unified V8, Spidermonkey and Chakra for all important platforms, but the maintenance seems suspended from April.

**Comment by ricardoquesada** [comment]
> We need a generic abstract layer of JSB API, now we are using SpiderMonkey API directly, but
> it's too unstable, APIs change more or less during each version update, this has affected :

yes yes yes.
In fact, once we have that API, we could use another engine if needed.
As far as I know, Apple opened the JavaScriptCore API, and it is framework now. Is that correct? If so, we could use it and there is no need to ship our own VM for iOS.

**Comment by pandamicro** [comment]
> In fact, once we have that API, we could use another engine if needed.
> As far as I know, Apple opened the JavaScriptCore API, and it is framework now. Is that correct? If so, we could use it and there is no need to ship our own VM for iOS.

Exactly, seems v8 is also smaller than spidermonkey on Android, so it's also beneficial for iOS/Android package size

Besides, as Apple opens JavaScriptCore API, I suppose they are gonna keep its backward compatibility, so maybe it's good to take JavaScriptCore API as our example.

**Comment by rampr** [comment]
Hi All,

I see spidermonkey was updated to v52 in here -> https://github.com/cocos-creator/cocos2d-x-lite/pull/675. Curious to understand why can't those changes be applied to cocos2d itself?

Thanks

**Comment by rampr** [comment]
Tagging @pandamicro and @minggo to understand your thoughts. Thank you.
